sound/soc/sof/ipc.c

```diff
@@ -639,90 +639,75 @@ EXPORT_SYMBOL(snd_sof_ipc_stream_posn);
 /*
  * IPC get()/set() for kcontrols.
  */
-
-int snd_sof_ipc_set_comp_data(struct snd_sof_ipc *ipc,
-			      struct snd_sof_control *scontrol, u32 ipc_cmd,
-			      enum sof_ipc_ctrl_type ctrl_type,
-			      enum sof_ipc_ctrl_cmd ctrl_cmd)
+int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
+				  struct snd_sof_control *scontrol,
+				  u32 ipc_cmd,
+				  enum sof_ipc_ctrl_type ctrl_type,
+				  enum sof_ipc_ctrl_cmd ctrl_cmd)
 {
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	int err;
+	u32 send_bytes;
+	int send;
+	int err = 0;
 
-	/* read firmware volume */
+	/* read or write firmware volume */
 	if (scontrol->readback_offset != 0) {
-		/* we can read value header via mmaped region */
-		snd_sof_dsp_block_write(sdev, sdev->mmio_bar,
-					scontrol->readback_offset, cdata->chanv,
-					sizeof(struct sof_ipc_ctrl_value_chan) *
-					cdata->num_elems);
-
-	} else {
-		/* write value via slower IPC */
-		cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
-		cdata->cmd = ctrl_cmd;
-		cdata->type = ctrl_type;
-		cdata->rhdr.hdr.size = scontrol->size;
-		cdata->comp_id = scontrol->comp_id;
-		cdata->num_elems = scontrol->num_channels;
-
-		/* send IPC to the DSP */
-		err = sof_ipc_tx_message(sdev->ipc,
-					 cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size,
-					 cdata, cdata->rhdr.hdr.size);
-		if (err < 0) {
-			dev_err(sdev->dev, "error: failed to set control %d values\n",
-				cdata->comp_id);
-			return err;
+		/* write/read value header via mmaped region */
+		switch (ctrl_type) {
+		case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+		case SOF_CTRL_TYPE_VALUE_COMP_GET:
+		case SOF_CTRL_TYPE_DATA_GET:
+			send = 0;
+			break;
+		case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		case SOF_CTRL_TYPE_DATA_SET:
+			send = 1;
+			break;
+		default:
+			return -EINVAL;
 		}
-	}
+		send_bytes = sizeof(struct sof_ipc_ctrl_value_chan) *
+		cdata->num_elems;
 
-	return 0;
-}
-EXPORT_SYMBOL(snd_sof_ipc_set_comp_data);
+		if (send)
```

Comment:
why not just move the switch-case here and remove the "send" variable completely?